wordpress-mobile/WordPress-iOS
# Remote Feature Flags: Separate local and remote feature flags

## Description

This PR proposes a new architecture for remote feature flags.

## Motivation

Currently, to retrieve the remote value for a feature flag, we have to do something similar to the following snippet:
```swift
let store = RemoteFeatureFlagStore()
let value = store.value(for: FeatureFlag.flagName)
```
This way is unintuitive. And could easily result in incorrect usage.

## Proposed Solution

Introduce a new `RemoteFeatureFlag` enum. The enum has an instance function `enabled()`, which fetches the remote value from the `RemoteFeatureFlagStore`, and falls back to the default value if a remote value isn't found. And it also checks if the flag is overridden. 

Usage would be as follows:
```swift
let value = RemoteFeatureFlag.flagName.enabled()
```
## Technical details

`FeatureFlag` is used in our apps and widgets, and extensions. To use anything inside `FeatureFlag`, it has to be added as a member to those extra targets. When I tried to add `RemoteFeatureFlagStore`, to the extra targets, I went into a rabbit hole of errors because of the networking code.

The proposed approach splits the remote feature flags into their own component. This way, we'll only have to add the networking layer to an app extension if the extension needs remote feature flags.

With this approach, the responsibility of `RemoteFeatureFlagStore` is stripped down to just retrieving the remote value. The previous responsibility of checking the override store and falling back to the default value is transferred to `RemoteFeatureFlag.enabled()`

Additionally, `RemoteFeatureFlag.enabled()` has an option for injecting a custom `RemoteFeatureFlagStore` and a custom `FeatureFlagOverrideStore`, which can be helpful for testing.

## Testing Instructions
1. Run the app
2. Go to the debug menu
3. Make sure the displayed Remote Feature Flag values match the `/mobile/feature-flags` endpoint
4. Enable a remotely disabled feature flag
5. Make sure the change affects the app
6. Disable a remotely enabled feature flag
7. Make sure the change affects the app

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:
